src/wasm/wasm-emscripten.cpp

```diff
@@ -702,45 +717,115 @@ void AsmConstWalker::noteNonLinear(Expression* curr) {
 
 void AsmConstWalker::visitLocalSet(LocalSet* curr) { sets[curr->index] = curr; }
 
+Const* AsmConstWalker::resolveConstAddr(Expression* arg, const Name& target) {
+  while (!arg->dynCast<Const>()) {
+    if (auto* get = arg->dynCast<LocalGet>()) {
+      // The argument may be a local.get, in which case, the last set in this
+      // basic block has the value.
+      auto* set = sets[get->index];
+      if (set) {
+        assert(set->index == get->index);
+        arg = set->value;
+      }
+    } else if (auto* value = arg->dynCast<Binary>()) {
+      // In the dynamic linking case the address of the string constant
+      // is the result of adding its offset to __memory_base.
+      // In this case are only looking for the offset with the data segment so
+      // the RHS of the addition is just what we want.
+      assert(value->op == AddInt32);
+      arg = value->right;
```

Comment:
Will we ever run into a chain of expressions that contain more than one binary expression? I'm wondering if we should be even more strict about what we accept, perhaps with a flag saying the next expression must be the final constant.

Reply:
I didn't write this code. It's the result of method extraction.